internal/controllers/machine/machine_controller_phases_test.go

```diff
@@ -1094,11 +1259,70 @@ func TestReconcileInfrastructure(t *testing.T) {
 					},
 				},
 			},
-			expectResult:  ctrl.Result{},
-			expectError:   false,
-			expectChanged: false,
+			InfraMachineGetError: nil,
+			expectResult:         ctrl.Result{},
+			expectError:          false,
 			expected: func(g *WithT, m *clusterv1.Machine) {
-				g.Expect(m.Status.InfrastructureReady).To(BeFalse())
+				g.Expect(m.Status.InfrastructureReady).To(BeTrue())
+				g.Expect(ptr.Deref(m.Spec.ProviderID, "")).To(Equal("test://id-1"))
+				g.Expect(ptr.Deref(m.Spec.FailureDomain, "")).To(Equal("foo"))
+				g.Expect(m.Status.Addresses).To(HaveLen(2))
+			},
+		},
+		{
+			name: "err reading infra machine when infrastructure have been ready (something different than not found), it should return error",
+			machine: &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "machine-test",
+					Namespace: metav1.NamespaceDefault,
+				},
+				Spec: clusterv1.MachineSpec{
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+						Kind:       "GenericInfrastructureMachine",
+						Name:       "infra-config1",
+					},
+				},
+				Status: clusterv1.MachineStatus{
+					InfrastructureReady: true,
+				},
+			},
+			infraMachine:         nil,
+			InfraMachineGetError: errors.New("some error"),
+			expectResult:         ctrl.Result{},
+			expectError:          true,
+			expected: func(g *WithT, m *clusterv1.Machine) {
+				g.Expect(m.Status.InfrastructureReady).To(BeTrue())
+				g.Expect(m.Status.FailureMessage).To(BeNil())
+				g.Expect(m.Status.FailureReason).To(BeNil())
+			},
+		},
+		{
+			name: "infra machine not found when infrastructure have been ready, should be treated as terminal error", // TODO: think about this. it will keep returning an error indefinitely, when the machine is already at a terminal state
```

Comment:
```suggestion
			name: "infra machine not found when infrastructure have been ready, should be treated as terminal error",
```

This will keep returning an error indefinitely, when the machine is already at a terminal state.
Is this correct?

Reply:
probably not correct, for a terminal state I'd expect the machine to not take up space in the queue and stay there until it's deleted

Reply:
Fixed using TerminalError from controller runtime which prevents the controller from going into backoff